sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -624,17 +654,42 @@ private WriteShardedBundlesToTempFiles(
       if (numShardsView != null) {
         shardingSideInputs.add(numShardsView);
       }
-      return input
-          .apply(
-              "ApplyShardingKey",
-              ParDo.of(new ApplyShardingKeyFn(numShardsView, destinationCoder))
-                  .withSideInputs(shardingSideInputs))
-          .setCoder(KvCoder.of(ShardedKeyCoder.of(VarIntCoder.of()), input.getCoder()))
-          .apply("GroupIntoShards", GroupByKey.create())
-          .apply(
-              "WriteShardsIntoTempFiles",
-              ParDo.of(new WriteShardsIntoTempFilesFn()).withSideInputs(getSideInputs()))
-          .setCoder(fileResultCoder);
+
+      if (getShardingFunction() == null) {
```

Comment:
Here, topology of pipeline is changing depending on available sharding function. In maling list @reuvenlax mentioned :
`
Initial thought on PR: we usually try to limit changing coders in these types of transforms to better support runners that allow in-place updates of pipelines. Can this be done without changing the coder?
`
Yes, switching sharding finctions (and having runner with restore / upgrade functionality) can lead to GBK state having different values in there which needs to be carefully decoded. @reuvenlax Is this what you were referring to?